core/domain/prod_validation_jobs_one_off.py

```diff
@@ -0,0 +1,156 @@
+    """
+    model = model_class.get_by_id(model_id)
+    return model is not None and model.deleted is False
+
+
+class BaseModelValidator(object):
+    """Base class for validating models."""
+
+    errors = {}
+
+    @classmethod
+    def _external_id_relationships(cls):
+        """Defines a mapping of external id to model class.
+
+        Returns a dictionary of which keys are field of the model to validate
+        and values are the corresponding model class. Checks the entries in that
+        field of the model should be valid non-deleted instances of the given
+        model classes.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _validate_external_id_relationships(cls, item):
+        """Check whether the external id properties on the model correspond
+        to valid instances.
+
+        Args:
+            item: UserSubscriptionsModel. Entity to validate.
+        """
+        multiple_models_keys_to_fetch = {}
+        for prop_name, model_class in (
+                cls._external_id_relationships().items()):
+            model_ids = getattr(item, prop_name)
+            if model_ids:
+                multiple_models_keys_to_fetch[prop_name] = (
+                    model_class, model_ids)
+        multiple_models = (
+            datastore_services.fetch_multiple_entities_by_ids_and_models(
+                multiple_models_keys_to_fetch.values()))
+
+        for prop_name, keys_to_fetch, model_class_models in zip(
+                multiple_models_keys_to_fetch.keys(),
+                multiple_models_keys_to_fetch.values(),
+                multiple_models):
+            model_class, model_ids = keys_to_fetch
+            for model_id, model in zip(model_ids, model_class_models):
+                if model is None or model.deleted:
+                    cls.errors['%s field check' % prop_name] = (
+                        'UserSubscriptionsModel id %s: based on field %s having'
+                        ' value %s, expect model %s with id %s but it doesn\'t'
+                        ' exist' % (
+                            item.id, prop_name, model_id,
+                            str(model_class.__name__), model_id))
+
+
+class UserSubscriptionsModelValidator(BaseModelValidator):
+    """Class for validating UserSubscriptionsModels."""
+
+    @classmethod
+    def _external_id_relationships(cls):
+        return {
+            'activity_ids': exp_models.ExplorationModel,
+            'collection_ids': collection_models.CollectionModel,
+            'general_feedback_thread_ids': (
+                feedback_models.GeneralFeedbackThreadModel),
+            'creator_ids': user_models.UserSettingsModel,
+        }
+
+    @classmethod
+    def validate(cls, item):
+        """Validates a UserSubscriptionsModel entity.
+
+        Args:
+            item: UserSubscriptionsModel. Entity to validate.
+        """
+        cls.errors.clear()
+
+        if not has_undeleted_model_for_id(
```

Comment:
I'm a little confused. What is this intended to check, exactly? I mean, if we already have the item, then why do we need to re-retrieve it and check its existence?

Reply:
Ahh, sorry for the confusion. this is to check there's a valid user for the UserSubscriptionsModel, by checking existence of UserSettingsModel.

Reply:
Ohh, I see, I misread, sorry! My bad.

That said, could we find a way to combine this check with the _validate_external_id_relationships() one? The type of check is the same ... check whether a given model class/ID appears in the datastore. If we could combine them, that will lead to fewer network calls to the datastore.

I think this might possibly entail a bit of a refactor (to pass error messages in with the models/ids to fetch), but let's try to streamline things properly for this one since we will be reusing the pattern in the future.

Reply:
Yes consolidating makes sense.